src/node/interfaces.cpp

```diff
@@ -942,9 +943,89 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(BlockWaitOptions options) override
+    {
+        // Delay calculating the current template fees, just in case a new block
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
+        const MillisecondsDouble tick{1000};
+
+        bool tip_changed{false};
+        do {
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                // Note that wait_until() checks the predicate before waiting
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    AssertLockHeld(notifications().m_tip_block_mutex);
+                    if (tip_changed) return true;
```

Comment:
@Sjors , does `tip_changed` ever become `true` at line 956 and we return from there? I think it will always be false hence can be deleted.

1. When we first enter this scope, the predicate is executed before blocking. If `tip_changed` is updated to `true`, it is returned. And when the predicate returned value is `true`, we do not wait; instead, we proceed to build and return a block template.

2. In subsequent executions of the predicate `tip_changed` will be `false` and we will check if the tip has changed, if yes `tip_changed` is updated to `true`, and returned, causing the wait to end. We then proceed to build and return a block template.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31283/files#r1960644463

Agree it looks like this line can be deleted. Also scope of `tip_changed` can be reduced so it is inside the do..while loop.

Reply:
Done both.